### PR TITLE
Fix TensorBoardLogger creating redundant experiment when finalizing

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Unblocks #12292, as I found a bug while helping out there.

The newly added test in test_tensorboard.py illustrates the issue: When nothing is logged, no experiment exists, but the call to finalize() would create a new experiment, then close it immediately again, leaving useless summary files behind.
This PR changes this to only flush/close or create directories when an experiment actually exists. This is consistent with all other loggers. 

### Does your PR introduce any breaking changes? If yes, please list them.

I don't think so. This was a peculiar edge case I can't believe anyone would need to rely on for a serious use case. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @awaelchli @edward-io @rohitgr7 @akihironitta

### CODE REVIEW

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py
Comment (awaelchli) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py:889]: the fact that this created version folders was just a result of the summary writer being created in finalize, and being immediately flushed with no (real) logs.

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/CHANGELOG.md
Comment (krshrimali) [inline src/pytorch_lightning/CHANGELOG.md:189]: _very_ minor :D

```suggestion
- Fixed an issue with `TensorBoardLogger.finalize` creating a new experiment when none was created during the Trainer's execution ([#14762](https://github.com/Lightning-AI/lightning/pull/14762))
```

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py
Comment (krshrimali) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py:883]: nit: was this left while debugging? 
```suggestion

```

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py
Comment (krshrimali) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py:888]: Thanks for adding a comment here! 🚀

File: tests/tests_pytorch/loggers/test_tensorboard.py
Comment (krshrimali) [inline tests/tests_pytorch/loggers/test_tensorboard.py:278]: Is there a use-case where user will also want to check if the `experiment` is None? Because in that case we wouldn't want them to use these protected objects/variables. If it's only for tests, then it's totally fine.

**Review by krshrimali — APPROVED** [review]
Thanks @awaelchli ! 2 minor nits (needed before merging I guess) and a small question (this question shouldn't block merging this PR), otherwise LGTM. 🚀

File: tests/tests_pytorch/loggers/test_tensorboard.py
Comment (awaelchli) [reply tests/tests_pytorch/loggers/test_tensorboard.py:278]: @krshrimali since this is a test, it fine to assert the internals here. Not all statements in a test does necessarily reflect how a user would interact with the object. 

The user would access logger.experiment or other public attributes.

Comment (krshrimali) [reply tests/tests_pytorch/loggers/test_tensorboard.py:278]: @awaelchli Sorry if I wasn't clear with my question, I understand it's okay to assert the internals in test case.

My question is - will a user ever want to check if the `experiment` is None? If yes, anyway they can do it without checking the internals? More of a question then a blocker review to the PR.

Comment (awaelchli) [reply tests/tests_pytorch/loggers/test_tensorboard.py:278]: I don't think they want to, and they shouldn't. They expect experiment not to be None. Which is why when the user calls logger.experiment, we create one if one does not exist (we create it lazily).

Comment (krshrimali) [reply tests/tests_pytorch/loggers/test_tensorboard.py:278]: Got it, thanks @awaelchli!

**Pull request merged**
